### DISCUSSION TITLE
help for frequency response based system identification

### DISCUSSION THREAD

**Discussion by Astik-2002** [baggepinnen/ControlSystemIdentification.jl#151] [General]
I'm trying to use this code to estimate a parametric transfer function from frequency response data `using ControlSystemIdentification
using ControlSystemsBase
using Plots
using LinearAlgebra
using CSV
using DataFrames

# Sample time
t = 0.01

# Read data from CSV file
data = CSV.read("/home/astik/Downloads/frequency_response_p.csv", DataFrame)

freq = data.Frequency
mag = data.Magnitude
phase = data.Phase

phase_radians = deg2rad.(phase)
# Convert magnitude and phase to complex form
response = mag .* exp.(1im .* phase_radians)

# Create FRD object
frd_data = FRD(response, freq)

G0 = tf(1.0, [1,1,1]) # Initial guess
G = tfest(frd_data, G0)`
And I'm getting this error ` MethodError: no method matching isless(::Int64, ::ComplexF64)

Closest candidates are:
  isless(::Missing, ::Any)
   @ Base missing.jl:87
  isless(::Any, ::Missing)
   @ Base missing.jl:88
  isless(::Integer, ::ForwardDiff.Dual{Ty}) where Ty
   @ ForwardDiff ~/.julia/packages/ForwardDiff/PcZ48/src/dual.jl:145
  ...

Stacktrace:
 [1] <(x::Int64, y::ComplexF64)
   @ Base ./operators.jl:352
 [2] >(x::ComplexF64, y::Int64)
   @ Base ./operators.jl:378
 [3] tfest(data::FRD{…}, p0::@NamedTuple{…}, link::Function; freq_weight::ComplexF64, refine::Bool, opt::Optim.BFGS{…}, opts::Optim.Options{…})
   @ ControlSystemIdentification ~/ControlSystemIdentification.jl/src/arx.jl:708
 [4] tfest(data::FRD{Vector{…}, Vector{…}}, p0::@NamedTuple{b::Vector{…}, a::Vector{…}}, link::Function)
   @ ControlSystemIdentification ~/ControlSystemIdentification.jl/src/arx.jl:683
 [5] tfest(::FRD{…}, ::TransferFunction{…}; kwargs::@Kwargs{})
   @ ControlSystemIdentification ~/ControlSystemIdentification.jl/src/arx.jl:748
 [6] tfest(::FRD{Vector{…}, Vector{…}}, ::TransferFunction{Continuous, ControlSystemsBase.SisoRational{…}})
   @ ControlSystemIdentification ~/ControlSystemIdentification.jl/src/arx.jl:744
 [7] top-level scope
   @ ~/ControlSystemIdentification.jl/examples/mat_to_frd.jl:26
Some type information was truncated. Use `show(err)` to see complete types.`

how to solve this?

**Comment by baggepinnen** [comment]
The problem is likely that you have written
```julia
frd_data = FRD(response, freq)
```
but the arguments should be reversed
```julia
frd_data = FRD(freq, response)
```

**Reply by Astik-2002** [reply]
Hey, thanks for this. it's working now.
I had another question related to the same. Suppose I know that the transfer function is going to be of the form k*s^2/(a4*s^4 + a3*s^3 + a2*s^2 + a1*s + a0), i.e with 4 poles and 2 zeros both at 0. how do I estimate the tf with this constraint, with the free variable being k (gain) and ai (i-> 0 to 4)?

**Reply by Astik-2002** [reply]
Also, to find transfer function in MATLAB non parametrically, we can use the tfestimate function. In this function, we can define the window, sample length, fft length and overlapping (for eg: tfestimate(input,output,[],512,1024,628)). how can we do the same here?

**Reply by Astik-2002** [reply]
I gave the tf([1.0, 0, 0],[1,1,1,1,1])) as the first guess (all coefficients as floats) to the tfest function, but the output tf didnot have the correct structure. when I used arx method with one numerator and four denominator coefficient, the tf had structure k*s^3 instead of k*s^2 in the numerator

**Reply by baggepinnen** [reply]
> Suppose I know that the transfer function is going to be of the form ks^2/(a4s^4 + a3s^3 + a2s^2 + a1*s + a0), i.e with 4 poles and 2 zeros both at 0. how do I estimate the tf with this constraint, with the free variable being k (gain) and ai (i-> 0 to 4)?

We do not yet support imposing such structure in `tfestimate`

> In this function, we can define the window, sample length, fft length and overlapping (for eg: tfestimate(input,output,[],512,1024,628)). how can we do the same here?

We use a different algorithm here, you control the width of the gaussian window using `σ` [here](https://baggepinnen.github.io/ControlSystemIdentification.jl/dev/freq/#ControlSystemIdentification.tfest).

> when I used arx method with one numerator and four denominator coefficient, the tf had structure ks^3 instead of ks^2 in the numerator

I don't understand what you mean, do you have an example? `arx` returns a discrete-time transfer function, so the Laplace variable is not present at all.
